### PR TITLE
support index tag

### PR DESCRIPTION
**Pull request by mfdeveloper508**
https://github.com/moov-io/iso8583/issues/281

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -481,74 +485,39 @@ func (m *Message) Unmarshal(v interface{}) error {
...
-		if _, set := m.fieldsMap[fieldIndex]; !set {
+		if _, set := m.fieldsMap[indexTag.Id]; !set {
 			continue
 		}
 
 		dataField := dataStruct.Field(i)
 		switch dataField.Kind() { //nolint:exhaustive
 		case reflect.Chan, reflect.Func, reflect.Map, reflect.Pointer, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
-			if dataField.IsNil() {
+			if dataField.IsNil() || dataField.IsZero() {
```
Comment (alovak) [inline message.go:507]: I'm curious, why do we check for zero value here?

File: exp_test.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
+package iso8583_test
```
Comment (alovak) [inline exp_test.go:1]: May I ask you to move the tests into `message_test` and to review if all test cases you may think about are covered by these tests?

File: exp_test.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+			{
+				name: "struct with *int type, no value and keepzero tag",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber *int   `index:"2,keepzero"`
```
Comment (alovak) [inline exp_test.go:147]: same here. I think we should create a new **int** and set the field value to `0`.

File: message.go [hunk 4]
```diff
@@ -481,74 +485,39 @@ func (m *Message) Unmarshal(v interface{}) error {
...
 		}
 
-		if _, set := m.fieldsMap[fieldIndex]; !set {
+		if _, set := m.fieldsMap[indexTag.Id]; !set {
 			continue
 		}
 
 		dataField := dataStruct.Field(i)
 		switch dataField.Kind() { //nolint:exhaustive
 		case reflect.Chan, reflect.Func, reflect.Map, reflect.Pointer, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
```
Comment (alovak) [inline message.go:506]: This comment a bit late and is not about the scope of this ticket, but about the previous PR. Looking at this line I think we should be specific about what do we support:
```suggestion
		case reflect.Pointer, reflect.Interface, reflect.Slice:
```

the same should be for other types (Int, Int64, String)
and finally, in the `default` we should return the error.

What do you think? We can create a separate ticket for this change.

File: exp_test.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+				packed, err := message.Pack()
+				require.NoError(t, err)
+
+				require.Equal(t, tt.expectedPackedString, string(packed))
+			})
+		}
+	})
+
+	t.Run("unpack", func(t *testing.T) {
```
Comment (alovak) [inline exp_test.go:169]: I would review `Unmarshal`code and write tests to cover all `if` and `case` blocks such as:
* test for setting []byte

and also test negative scenarios:
* using map
* use slice of strings
* etc.

File: message_test.go [hunk 6]
```diff
@@ -1415,21 +1415,17 @@ func Test_getFieldIndex(t *testing.T) {
...
 
-		index, err := getFieldIndex(st.Type().Field(0))
-
-		require.NoError(t, err)
-		require.Equal(t, 1, index)
+		indexTag := field.NewIndexTag(st.Type().Field(0))
+		require.Equal(t, 1, indexTag.Id)
 	})
 
 	t.Run("returns index from field tag instead of field name when both match", func(t *testing.T) {
```
Comment (alovak) [inline message_test.go:1422]: as we now have `field/index_tag.go` I think we should move all related tests into `field/index_tag_test.go` and remove them from other places. In index tag test we can test different test cases as well using table tests.
